File: searchnsort/find_3_largest.cpp
//AlgoExpert - find largest 3 numbers

#include <vector>
#include<climits>
using namespace std;

void updateLargest(vector<int> &threeLargest, int num);
void shiftAndUpdate(vector<int> &largest, int num, int idx);

vector<int> findThreeLargestNumbers(vector<int> array){
	vector<int> threelargest{INT_MIN, INT_MIN, INT_MIN};
	for (num : array){
		updatelargest(threelargest, num);
	}
	return threelargest;
}

void updatelargest(vector<int> &threelargest,int num){
	if (num > threelargest[2]){
		shiftandupdate(threelargest, num, 2);
	}
	elif (num > threelargest[1]){
		shiftandupdate(threelargest, num, 1);
	}
	elif (num > threelargest[0]){
		shiftandupdate(threelargest, num, 0);
	}
}

void shiftandupdate(vector<int> &array,int num,int idx){
	for(int i = 0; i <= idx; i++){
		if (i == idx){
			array[i] = num;
		}
		else{
			array[i] = array[i + 1];
		}
	}
}

File: searchnsort/find_3_largest.py
'''
AlgoExpert - find largest 3 numbers
'''
def findThreeLargestNumbers(array):
	threelargest = [None, None, None]
	for num in array:
		updatelargest(threelargest, num)
	return threelargest

def updatelargest(threelargest, num):
	if threelargest[2] is None or num > threelargest[2]:
		shiftandupdate(threelargest, num, 2)
	elif threelargest[1] is None or num > threelargest[1]:
		shiftandupdate(threelargest, num, 1)
	elif threelargest[0] is None or num > threelargest[0]:
		shiftandupdate(threelargest, num, 0)

def shiftandupdate(array, num, idx):
	for i in range(idx + 1):
		if i == idx:
			array[i] = num
		else:
			array[i] = array[i + 1]

